Added app.extend()

boycce opened the pull request:
Added `app.extend()`, which provides an easy way to update certain properties of a setting.
Example:
```
app.set('foo', { 'bar' : 1 })
app.extend('foo', { 'baz' : 1 });
app.get('foo');
// => { "bar" : 1, "baz" : 1 }
```

Related issue #3499

boycce committed:
Added app.extend()

Added app.extend(), which provides an easy way to update certain properties of a setting.

wesleytodd commented:
Hey @Boycce.  This is an interesting addition.  Can you provide some real world examples of why you would like this?

Could you also take a look at how you would do this addition considering this planned change: https://github.com/expressjs/express/pull/3218

boycce commented:
Hey @wesleytodd I see no issues migrating `.extend()` along with the other getter/setter methods. 

One example would be if you needed to update stored configurations for a particular service while keeping to express's familiar set/get/expand* methods:
```
app.set(braintree, {
  'environment': 'production',
  'merchantId': 'os9ksSioSeyH87fhd',
  'publicKey': 'kcYsjcSlkfd8Hs2jd',
  'privateKey': '1z844211b3d012228dee47279ef30000',
  'merchantAccountIdNZD': 'userNZD',
  'merchantAccountIdUSD': 'userUSD'
})

app.expand('braintree', {
  'privateKey': '3syd7D9l02hh11dB91p02kdufhnw712a'
})
```
Maybe @brunoscopelliti could also share his use case for needing `extend()`.

brunoscopelliti commented:
In some cases I found useful to set related fields under the same object.
As example:

```js
app.set("log_level", "verbose");
app.set("log_active", true);

// vs 

app.set("log_config", { level: "verbose", active: true });
```

When I use the latter approach I would like to have a better method to update a single field, than:

```js
app.set("log_config", Object.assign({}, app.get("log_config"), { dir: "/db/log/" }));
```

wesleytodd commented:
Hey @Boycce, first off, I hope those are not your production configs ;).  But second @Boycce & @brunoscopelliti, your use cases seem reasonable.  I prefer `extend` over `expand`, assuming that is not just a typo.

I think this change is something which could be landed on `4.x` as well as the `5.x` branch.  So can you update this PR to point to `master` for `4.x`, then create a PR to [`store-settings`](https://github.com/wesleytodd/store-settings) so that I can update #3218 with this?

dougwilson commented:
Is this handling inherited settings correctly? It seems like it's actually altering the parent property in addition to setting the child property to the same object instance. I feel like the parent object should not be getting modified, so the setting on the parent app should remain the same when the setting on the child app is extended.

wesleytodd commented:
I dont think it would set the parent's property because it is prototypicaly inherited, not done by reference.  So if it was initially from the parent, it would just result in settings it's own property with a fully copy of the parent and not touching the parent values.  I could be wrong though, would be worth having a test around this.

See: https://github.com/expressjs/express/pull/3568/files#diff-5372f626ee15242f1e2c6eb31655b4faR116

dougwilson commented:
I did test this and it is altering the parent property _as well as_ setting the child property to the same object reference.

wesleytodd commented:
Ahh, ok then I miss-read how this works, sorry.  Yes I agree it *should not modify* the parent.

dougwilson commented:
Here is the test I ran. I feel like this test should pass, but it currently does not:
```js
var assert = require('assert')
var express = require('express')

var app1 = express()
var app2 = express()

app1.set('log_config', { level: "verbose", active: false })

assert.equal(app1.get('log_config').active, false)
assert.equal(app2.get('log_config'), undefined)

app1.use(app2)

assert.equal(app1.get('log_config').active, false)
assert.equal(app2.get('log_config').active, false)

app2.extend('log_config', { active: true })

assert.deepEqual(app1.get('log_config'), { level: "verbose", active: false })
assert.deepEqual(app2.get('log_config'), { level: "verbose", active: true })

console.log('success')
```

```
assert.js:81
  throw new assert.AssertionError({
  ^
AssertionError: { level: 'verbose', active: true } deepEqual { level: 'verbose', active: false }
    at Object.<anonymous> (test.js:19:8)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.runMain (module.js:604:10)
    at run (bootstrap_node.js:389:7)
    at startup (bootstrap_node.js:149:9)
    at bootstrap_node.js:504:3
```

wesleytodd commented on lib/application.js:
```diff
@@ -377,6 +377,50 @@ app.set = function set(setting, val) {
+  }
+
+  // extend value
+  this.settings[setting] = Object.assign(this.settings[setting], val);
```
The first argument in `Object.assign` is the destination object.  So now that I read this more closely, this is probably the culprit for the parent issue, not to mention it being redundant to set the value twice.  Start by just doing this:

```
this.settings[setting] = Object.assign({}, currentSetting, val)
```

boycce replied:
Seems to be passing Douglas's reference tests. Depending on `currentSetting` I will also let the base object be a variable of either an array or literal.

boycce committed:
Shouldn't extend by reference

`extend` shouldn't be setting the propertry to the same object reference.

boycce commented:
Updated to 4.x and removed object referencing.
Would you like me to support Node 3.3 and below concerning `Object.assign()`

wesleytodd commented on lib/application.js:
```diff
@@ -383,6 +383,53 @@ app.set = function set(setting, val) {
+  }
+
+  // extend from the intial object type
+  var newObject = Array.isArray(this.settings[setting]) ? [] : {};
```
I believe that by adding this `isArray` check you loose support for node `0.12`, making this unable to release on `4.x`.  Also using `Object.assign` on an array is VERY miss leading to consumers.  If I were you I would make this method ONLY work on plain objects and throw like you did above on line 411.

wesleytodd commented:
I left a comment about the `isArray` usage, but the same does apply to `Object.assign`.

boycce committed:
Support for previous Node versions and now only accepts objects

Added `utils.extendObject()` to support previous Node versons
Removed support for extending arrays and throws an error

dougwilson commented on lib/utils.js:
```diff
@@ -262,6 +262,31 @@ exports.setCharset = function setCharset(type, charset) {
+ * @private
+ */
+
+exports.extendObject = function(target, source) {
```
Instead of adding more code here, can the dependency `utils-merge` we already have perform this function? Or if not, what's the difference between this code and `utils-merge`?

boycce replied:
Hi @dougwilson, sorry I didn't see this merge util, I was scanning for an extending `extend()`  function or corresponding library 🤐. This passes my tests:
```
this.settings[setting] = merge(merge({}, this.settings[setting]), val)
```

wesleytodd replied:
Ugh, I didn't realize that `utils-merge` did not support variadic calls.  I think this is, sadly, the best way at the moment, but I think we should add to the roadmap "removing utils-merge in favor of object.assign or an equivilant fallback".

dougwilson commented:
What do you think we should do for the case where `.extend` is called on a setting that is already set but is not an object? Right now it is silently overwritten. Do we think that is the correct thing to do?

Example:
```
$ node -pe 'require("express")().set("thing","foo").extend("thing",{"foo":"bar"}).get("thing")'
{ foo: 'bar' }
```

dougwilson commented on lib/application.js:
```diff
@@ -383,6 +384,50 @@ app.set = function set(setting, val) {
+ *
+ *    app.set('foo', { 'bar' : 1 })
+ *    app.expand('foo', { 'baz' : 1 });
+ *    app.expand('foo');
```
@wesleytodd thoughts on this? I'm not sure what the use of having `.extend` called with one argument returning the current value -- that's already covered with `.get` and `.set` I don't think we need a third method to get the value of a setting, right?

wesleytodd replied:
Agreed.  Not sure if it should error, but it shouldn't do anything and shouldn't be documented.

dougwilson replied:
May be useful to throw error that required argument is missing.

boycce replied:
OK, I will edit this later today.

dougwilson commented on lib/application.js:
```diff
@@ -383,6 +384,50 @@ app.set = function set(setting, val) {
+  if (typeof val === 'undefined') {
+    return this;
+  } else if (val === null || typeof val !== 'object' || val instanceof Array) {
+    throw new Error('value needs to be an object');
```
This should probably be a `TypeError`

dougwilson commented on lib/application.js:
```diff
@@ -383,6 +384,50 @@ app.set = function set(setting, val) {
+
+  if (typeof val === 'undefined') {
+    return this;
+  } else if (val === null || typeof val !== 'object' || val instanceof Array) {
```
I would recommend `Array.isArray` instead of `instanceof Array`. All the `instanceof Array`s were removed in Express 1.0.

wesleytodd replied:
When looking at node.green, I think in my other comment I mentioned that I thought that wasn't supported in node `0.12`, but I think I looked at the wrong field, and it has been so long since I have used `0.12` I couldn't remember what was supported...

http://node.green/#ES2015-subclassing-Array-is-subclassable-Array-isArray-support

wesleytodd commented:
> What do you think we should do for the case where .extend is called on a setting that is already set but is not an object

I think it should refuse to work on any property which is not a plain object.

dougwilson commented:
> I think it should refuse to work on any property which is not a plain object.

Ah, that is a good point. If a setting is an instance, calling extend here would remove the original prototype, so yea, we don't want to destroy the object. We can always add additional things to extend as semver minors so just only working on plain objects sounds like a good first implementation.

boycce commented:
@dougwilson, would you like me to add a `isPlainObject` export to the utils file, or require it from the [lodash module](https://github.com/lodash/lodash/tree/4.0.6-npm-packages/lodash.isplainobject) in regards to checking if the object is in fact a plain object from the Object constructor.

I could use something like this:
``` javascript
function isPlainObject(obj) {
  return  typeof obj === 'object'
    && obj !== null
    && obj.constructor === Object
    && Object.prototype.toString.call(obj) === '[object Object]'
}
```

wesleytodd commented:
My point was more along the lines of, "don't try to merge arrays" and "don't try to handle complex types".  In re-reading it I see where this led, and I think I was making it too complicated.

So adding a dep for this is out of the question IMO.  And there is a bunch of valid use cases for merging "object like" things, so the full `isPlainObject` is also not a good path forward.  I would think just doing this is the best:

```javascript
app.extend = function(setting, val) {
  if (val === null || typeof val !== 'object' || Array.isArray(val)) {
    throw new Error('value needs to be an object');
  }

  debug('extend "%s" with %o', setting, val);

  var currentSetting = this.settings[setting];

  if (typeof currentSetting === 'undefined') {
    this.settings[setting] = val;
  } else {
    this.settings[setting] = extendObject({}, currentSetting, val);
  }

  return this;
};
```

Thoughts?

boycce commented:
Sounds good by me.

dougwilson commented:
So, this pull request is a bit stale, and looking back, it doesn't seem like there was really a consensus about adding this to core express. As pointed out in the conversation above, there are certainly different extending strategies one can make... And with things like the prototype pollution issues going around and the fact that, at it's core, it's not really much more than `app.set('foo', Object.assign(app.get('foo') || {}, { foo: 'bar' }))` but without the flexibility to perform your own merge strategies, it just seems more complexity into core than it's worth...
